### PR TITLE
crimson/osd: OpSequencer allows for out-of-order ClientRequests

### PR DESCRIPTION
**Pull request by rzarzynski**
This is basically a RFC. Tested locally just with `rados bench`; thrashing tests at Sepia are TBD.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/osd/osd_operation_sequencer.h [hunk 1]
```diff
@@ -69,10 +69,40 @@ class OpSequencer {
...
+         ++it) {
+      // we're iterating over the operation registry of all ClientRequests.
+      // this list aggrates every single instance in the system, and thus
+      // we need to skip operations coming from different client's session
+      // or targeting different PG. The op we're interested in should be
+      // somewhere between the operation that early exists (this_op) and
+      // the most recently registered one.
+      // as the out-of-order exit is supposed to happen on cold paths only,
+      // the extra overhead is a thing we can live with.
+      auto* maybe_next_op = std::addressof(dynamic_cast<ClientRequest&>(*it));
```
Comment (tchaikov) [inline src/crimson/osd/osd_operation_sequencer.h:89]: why `dynamic_cast`? can we just use `static_cast` instead? i am in favor of not using RTTI if we don't have to.

File: src/crimson/osd/osd_operation_sequencer.h [hunk 2]
```diff
@@ -31,11 +31,11 @@ class OpSequencer {
             typename Result = std::invoke_result_t<FuncT>>
   seastar::futurize_t<Result>
   start_op(HandleT& handle,
-           uint64_t prev_op,
-           uint64_t this_op,
+           const Operation* const prev_op,
```
Comment (tchaikov) [inline src/crimson/osd/osd_operation_sequencer.h]: might want to expand a little bit on the rationales behind this change in the commit message.

File: src/crimson/osd/osd_operation_sequencer.h [hunk 3]
```diff
@@ -69,10 +69,40 @@ class OpSequencer {
...
+         it != OperationRegistryI::op_list::s_iterator_to(*last_issued);
+         ++it) {
+      // we're iterating over the operation registry of all ClientRequests.
+      // this list aggrates every single instance in the system, and thus
+      // we need to skip operations coming from different client's session
+      // or targeting different PG. The op we're interested in should be
+      // somewhere between the operation that early exists (this_op) and
+      // the most recently registered one.
+      // as the out-of-order exit is supposed to happen on cold paths only,
+      // the extra overhead is a thing we can live with.
```
Comment (xxhdx1985126) [inline src/crimson/osd/osd_operation_sequencer.h:88]: Could you elaborate a little bit about the "out-of-order exit"?

File: src/crimson/osd/osd_operation_sequencer.h [hunk 4]
```diff
@@ -69,10 +69,40 @@ class OpSequencer {
...
   }
+  void finish_op_out_of_order(ClientRequest* const this_op) {
+    assert(this_op);
+    for (auto it = OperationRegistryI::op_list::s_iterator_to(*this_op);
+         it != OperationRegistryI::op_list::s_iterator_to(*last_issued);
+         ++it) {
+      // we're iterating over the operation registry of all ClientRequests.
+      // this list aggrates every single instance in the system, and thus
+      // we need to skip operations coming from different client's session
+      // or targeting different PG. The op we're interested in should be
```
Comment (xxhdx1985126) [inline src/crimson/osd/osd_operation_sequencer.h:84]: I think the op sequencer is already unique for each connection/pg pair, which, I think, makes the op sequencer allow for out-of-order op execution from different connection/pg pair. Am I missing anything?

**Comment by athanatos** [comment]
The main change here seems to be using pointers within the ops so that we can correct the prev id/pointer in an early exit, I think?  The unblocked cv broadcast/check seems like a relatively expensive way to identify the next op to run (we'll tend to wake them all?) It seems like we should be able to do this more cleanly by embedding a linked list for each session/client pair -- that would also simplify removing an early exit op.

File: src/crimson/osd/osd_operation_sequencer.h (hunk 3 above)
Comment (xxhdx1985126) [reply src/crimson/osd/osd_operation_sequencer.h:88]: IIUC, "out-of-order exit" can only happen when the op is read-only, am I right?

**Comment by tchaikov** [comment]
> The main change here seems to be using pointers within the ops so that we can correct the prev id/pointer in an early exit, I think? 

yes.

> The unblocked cv broadcast/check seems like a relatively expensive way to identify the next op to run (we'll tend to wake them all?) It seems like we should be able to do this more cleanly by embedding a linked list for each session/client pair -- that would also simplify removing an early exit op.

presumably, the pg interval changes while performing client requests should not happen a lot, i think.

**Comment by athanatos** [comment]
> > The unblocked cv broadcast/check seems like a relatively expensive way to identify the next op to run (we'll tend to wake them all?) It seems like we should be able to do this more cleanly by embedding a linked list for each session/client pair -- that would also simplify removing an early exit op.
> 
> presumably, the pg interval changes while performing client requests should not happen a lot, i think.

The unblocked cv broadcast/check happens to every op while waiting on the op in the next stage, I think.  It's not specific to pg interval changes.  Of course, this patch doesn't *add* that property -- it's already there -- but adding an embedded linked list to avoid it would I think also simplify the pg interval change case.

File: src/crimson/osd/osd_operation_sequencer.h (hunk 3 above)
Comment (rzarzynski) [reply src/crimson/osd/osd_operation_sequencer.h:88]: Nope. Before the patches `OpSequencer` was assuming every single (inside a specific session / pg boundaries) finishes in the same exact order as it was entered. This isn't truly the case as we have a bunch of early exists like `send_incremenetal_map()` on misdirected op. See https://github.com/ceph/ceph/pull/41866.

File: src/crimson/osd/osd_operation_sequencer.h (hunk 1 above)
Comment (rzarzynski) [reply src/crimson/osd/osd_operation_sequencer.h:89]: Conceptually it acts like an assertion for the assumption I made. Though, not having strong feelings towards it and can change to `static_cast`.

**Comment by rzarzynski** [comment]
@athanatos: my impression is the check is never performed outside the _interval change period_. It's because `last_unblocked` is updated very early, **before** returning the future from the passed callable:

```cpp
      auto result = seastar::futurize_invoke(std::move(do_op));
      // unblock the next one
      last_unblocked = this_op;
      unblocked.broadcast();
      return result;
```

The question is how costly is the `broadcast()` when nobody waits.

File: src/crimson/osd/osd_operation_sequencer.h (hunk 2 above)
Comment (rzarzynski) [reply src/crimson/osd/osd_operation_sequencer.h]: Sure.

**Review by athanatos — APPROVED** [review]
Ok, LGTM.

File: src/crimson/osd/osd_operation_sequencer.h (hunk 1 above)
Comment (tchaikov) [reply src/crimson/osd/osd_operation_sequencer.h:89]: @rzarzynski how about using `static_cast<>()`. but also have
```c++
assert(dynamic_cast<ClientRequest&>(*it));
```
?

**Comment by rzarzynski** [comment]
Closing in favor of #43011.

**Pull request closed without merging**